### ISSUE TITLE
[BUG] batch_params example not running with torch interface

### ISSUE THREAD

**Issue by resting-dove** [PennyLaneAI/pennylane#2098]
### Expected behavior

[https://pennylane.readthedocs.io/en/stable/code/api/pennylane.batch_params.html](url) gives an example for the use of `qml.batch_params` to use batched inputs for QNodes. The website warns that not everything has been updated to support this feature and asks for a Githun Issue, when running into such a problem. 

This issue is potentially already covered in #2037, but encountered in a more explicit and basic form here. 

This is the example code from the website that works:

```python
import pennylane as qml
from pennylane import numpy as np
import torch


dev = qml.device("default.qubit", wires=3)

@qml.batch_params
@qml.beta.qnode(dev)
def circuit(x, weights):
    qml.RX(x, wires=0)
    qml.RY(0.2, wires=1)
    qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1, 2])
    return qml.expval(qml.Hadamard(0))

batch_size = 3
x = np.linspace(0.1, 0.5, batch_size)
rng = np.random.default_rng(seed=1234)
weights = rng.random((batch_size, 10, 3, 3), requires_grad=True)

circuit(x, weights)
```
`>>> tensor([ 0.00800498,  0.2735391 , -0.24395442], requires_grad=True)`

### Actual behavior

When using `interface="torch"` and `torch.tensor` inputs to the QNode an error is raised.

### Source code

```shell
import pennylane as qml
from pennylane import numpy as np
import torch

dev = qml.device("default.qubit", wires=3)

@qml.batch_params
@qml.beta.qnode(dev, interface="torch")
def circuit(x, weights):
    qml.RX(x, wires=0)
    qml.RY(0.2, wires=1)
    qml.templates.StronglyEntanglingLayers(weights, wires=[0, 1, 2])
    return qml.expval(qml.Hadamard(0))

batch_size = 3
x = np.linspace(0.1, 0.5, batch_size)
rng = np.random.default_rng(seed=1234)
weights = rng.random((batch_size, 10, 3, 3), requires_grad=True)

circuit(torch.tensor(x), torch.tensor(weights))
```

### Tracebacks

```shell
---------------------------------------------------------------------------
IndexError                                Traceback (most recent call last)
/opt/conda/lib/python3.8/site-packages/pennylane/transforms/batch_params.py in batch_params(tape, all_operations)
    115     try:
--> 116         batch_dim = qml.math.shape(params[0])[0]
    117     except IndexError:

IndexError: list index out of range

During handling of the above exception, another exception occurred:

IndexError                                Traceback (most recent call last)
/tmp/ipykernel_179/1981642925.py in <module>
     14 weights = rng.random((batch_size, 10, 3, 3), requires_grad=True)
     15 
---> 16 circuit(torch.tensor(x), torch.tensor(weights))

/opt/conda/lib/python3.8/site-packages/pennylane/transforms/batch_transform.py in _wrapper(*args, **kwargs)
    287             shots = kwargs.pop("shots", False)
    288             qnode.construct(args, kwargs)
--> 289             tapes, processing_fn = self.construct(qnode.qtape, *targs, **tkwargs)
    290 
    291             interface = qnode.interface

/opt/conda/lib/python3.8/site-packages/pennylane/transforms/batch_transform.py in construct(self, tape, *args, **kwargs)
    407             tape = self.expand_fn(tape, *args, **kwargs)
    408 
--> 409         tapes, processing_fn = self.transform_fn(tape, *args, **kwargs)
    410 
    411         if processing_fn is None:

/opt/conda/lib/python3.8/site-packages/pennylane/transforms/batch_params.py in batch_params(tape, all_operations)
    116         batch_dim = qml.math.shape(params[0])[0]
    117     except IndexError:
--> 118         raise ValueError(f"Parameter {params[0]} does not contain a batch dimension.") from None
    119 
    120     unbatched_params = [[] for i in range(batch_dim)]

IndexError: list index out of range
```

### System information

```shell
Name: PennyLane
Version: 0.19.1
Summary: PennyLane is a Python quantum machine learning library by Xanadu Inc.
Home-page: https://github.com/XanaduAI/pennylane
Author: None
Author-email: None
License: Apache License 2.0
Location: /opt/conda/lib/python3.8/site-packages
Requires: semantic-version, networkx, pennylane-lightning, appdirs, toml, scipy, cachetools, numpy, autograd, autoray
Required-by: pennylane-qulacs, PennyLane-Lightning
Platform info:           Linux-4.15.0-158-generic-x86_64-with-glibc2.10
Python version:          3.8.8
Numpy version:           1.19.2
Scipy version:           1.6.1
Installed devices:
- default.gaussian (PennyLane-0.19.1)
- default.mixed (PennyLane-0.19.1)
- default.qubit (PennyLane-0.19.1)
- default.qubit.autograd (PennyLane-0.19.1)
- default.qubit.jax (PennyLane-0.19.1)
- default.qubit.tf (PennyLane-0.19.1)
- default.qubit.torch (PennyLane-0.19.1)
- default.tensor (PennyLane-0.19.1)
- default.tensor.tf (PennyLane-0.19.1)
- qulacs.simulator (pennylane-qulacs-0.16.0)
- lightning.qubit (PennyLane-Lightning-0.19.0)

​PyTorch version: 1.8.1
Is debug build: False
CUDA used to build PyTorch: 11.1
ROCM used to build PyTorch: N/A

OS: Ubuntu 18.04.5 LTS (x86_64)
GCC version: (Ubuntu 7.5.0-3ubuntu1~18.04) 7.5.0
Clang version: Could not collect
CMake version: version 3.10.2
Libc version: glibc-2.27

Python version: 3.8.8 (default, Feb 24 2021, 21:46:12)  [GCC 7.3.0] (64-bit runtime)
Python platform: Linux-4.15.0-158-generic-x86_64-with-glibc2.10
Is CUDA available: True
CUDA runtime version: 11.1.105
GPU models and configuration: GPU 0: NVIDIA GeForce GTX 1080 Ti
Nvidia driver version: 470.57.02
cuDNN version: Probably one of the following:
/usr/lib/x86_64-linux-gnu/libcudnn.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_adv_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_cnn_train.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_infer.so.8.0.5
/usr/lib/x86_64-linux-gnu/libcudnn_ops_train.so.8.0.5
HIP runtime version: N/A
MIOpen runtime version: N/A

Versions of relevant libraries:
[pip3] numpy==1.19.2
[pip3] torch==1.8.1
[pip3] torchelastic==0.2.2
[pip3] torchtext==0.9.1
[pip3] torchvision==0.9.1
[conda] blas                      1.0                         mkl
[conda] cudatoolkit               11.1.74              h6bb024c_0    nvidia
[conda] ffmpeg                    4.3                  hf484d3e_0    pytorch
[conda] mkl                       2020.2                      256
[conda] mkl-service               2.3.0            py38he904b0f_0
[conda] mkl_fft                   1.3.0            py38h54f3939_0
[conda] mkl_random                1.1.1            py38h0573a6f_0
[conda] numpy                     1.19.2           py38h54aff64_0
[conda] numpy-base                1.19.2           py38hfa32c7d_0
[conda] pytorch                   1.8.1           py3.8_cuda11.1_cudnn8.0.5_0    pytorch
[conda] torchelastic              0.2.2                    pypi_0    pypi
[conda] torchtext                 0.9.1                      py38    pytorch
[conda] torchvision               0.9.1                py38_cu111    pytorch
```

### Existing GitHub issues

- [X] I have searched existing GitHub issues to make sure the issue does not already exist.

**Comment by josh146** [comment]
Thanks for reporting this @resting-dove! This is super useful info to have to help improve this feature.

Currently, the default behaviour of `@batch_params` is to batch all parameters that are _trainable_. So if you explicitly mark both Torch tensors as trainable,

```pycon
>>> x = torch.tensor(x, requires_grad=True)
>>> weights = torch.tensor(weights, requires_grad=True)
>>> circuit(x, weights)
tensor([ 0.0080,  0.2735, -0.2440], dtype=torch.float64,
       grad_fn=<SqueezeBackward0>)
```

then this will work. However, if one is not trainable and the other is, issues will occur:

```pycon
>>> x = torch.tensor(x, requires_grad=False)
>>> weights = torch.tensor(weights, requires_grad=True)
>>> circuit(x, weights)
  File "/home/josh/miniconda3/lib/python3.8/site-packages/autoray/autoray.py", line 409, in translated_function
    return fn(**new_kwargs)
RuntimeError: stack expects each tensor to be equal size, but got [] at entry 0 and [3] at entry 1
```

I believe this is the same issue that is being discussed in #2037

**Comment by resting-dove** [comment]
Thanks for the quick reply @josh146!
